django/db/models/fields/related.py

```diff
@@ -1885,6 +1899,30 @@ def __init__(self, to, to_field=None, related_name=None, related_query_name=None
             # be correct until contribute_to_class is called. Refs #12190.
             to_field = to_field or (to._meta.pk and to._meta.pk.name)
 
+        if on_delete is None:
+            warnings.warn(
+                "on_delete will be a required kwarg for %s in Django 2.0. "
+                "Set it to models.CASCADE if you want to maintain the current default behavior. "
+                "See https://docs.djangoproject.com/en/%s/ref/models/fields/"
+                "#django.db.models.ForeignKey.on_delete" % (
+                    self.__class__.__name__,
+                    get_docs_version(),
+                ),
+                RemovedInDjango20Warning, 2)
+            on_delete = CASCADE
+
+        if not callable(on_delete) and callable(to_field):
```

Comment:
I think the best we can do here is just `if not callable(on_delete)`, which will catch a case like `ForeignKey(SomeModel, 'somefield')`. I don't think there's anything we can do about a case like `ForeignKey(SomeModel, 'somefield', on_delete=models.PROTECT)`, because as @timgraham points out, that will raise a `TypeError` immediately

So I guess that either means we break that case immediately, or we do a longer two-step deprecation (first deprecate positional specification of `to_field`, then later swap the position of the arguments), or we revert to @timgraham's initial suggestion of just leaving the argument order alone and having `on_delete` permanently be a required keyword argument.

I'm feeling pretty muddy on the topic at this point. Anyone else have a clear sense of what we should do?